### PR TITLE
Add `module_ctx.extension_metadata`

### LINKED ISSUE #17908: Implement "Automatic use_repo fixups for module extensions"
Proposal: https://docs.google.com/document/d/1dj8SN5L6nwhNOufNqjBhYkk5f-BJI_FPYWKxlB3GAmA/edit

### PR DESCRIPTION
**Pull request by fmeum**
Fixes #17908

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionContext.java [hunk 1]
```diff
@@ -117,4 +120,54 @@ public StarlarkList<StarlarkBazelModule> getModules() {
...
+      parameters = {
+          @Param(
+              name = "root_module_direct_deps",
+              doc = "foo",
+              positional = false,
+              named = true,
+              defaultValue = "None",
+              allowedTypes = {
+                  @ParamType(type = Sequence.class, generic1 = String.class),
+                  @ParamType(type = NoneType.class)}),
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionContext.java:136]: What do you think about having the string "all" as a possible value for this param to replace `module_ctx.use_all_repos` (similarly for dev deps)?

The obvious downside is the weird typing, but the advantage of this is that we remove some mental burden and _if_ we were to have other fields on ExtensionMetadata later on, it's more composable.

Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionContext.java:136]: yeah. now the question is where to put this as_module_extension thing :)

also notable that use_all_repos is implementable in terms of the "all" thing, but not vice versa. Mostly due to the composability point.

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionContext.java:137]: Result of an offline discussion: We will go for `"all"`, but realize a wrapper for the simple use case of a WORKSPACE macro either in skylib or Bazel itself.

**Comment by fmeum** [comment]
@Wyverald I pushed a new version that supports the `"all"` syntax. Could you take another look? I would add more tests afterwards.

**Review by Wyverald — COMMENTED** [review]
looks good overall!

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionMetadata.java [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+            throw Starlark.errorf("root_module_direct_deps contained the following repositories "
+                + "not generated by the extension: %s", String.join(", ", invalidRepos));
+          }
+        }
+        return Optional.ofNullable(explicitRootModuleDirectDeps);
+      case REGULAR:
+        return Optional.of(ImmutableSet.copyOf(allRepos));
+      case DEV:
+        return Optional.of(ImmutableSet.of());
+      default:
```
[the remark this answers is not in the document]
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionMetadata.java:255]: It is still needed for enums with `switch` _statements_. Unfortunately switch _expressions_ aren't available at Bazel's Java language level yet.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionMetadata.java [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+      message += String.format(
+          "Not imported, but declared as direct dependencies (may cause the build to fail):\n    %s\n\n",
+          String.join(", ", missingImports));
+    }
+
+    var indirectDepImports = ImmutableSortedSet.copyOf(
+        Sets.difference(Sets.intersection(allActualImports, allRepos), allExpectedImports));
+    if (!indirectDepImports.isEmpty()) {
+      message += String.format(
+          "Imported, but not declared as direct dependencies:\n    %s\n\n",
```
[the remark this answers is not in the document]
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionMetadata.java]: I reworded the messages to hopefully make this more clear.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java [hunk 4]
```diff
@@ -48,14 +50,34 @@ public abstract class ModuleExtensionUsage {
...
+   * a dev dependency.
+   */
+  public abstract ImmutableSet<String> getDevImports();
+
+  /**
+   * All the tags specified by this module for this extension.
+   */
   public abstract ImmutableList<Tag> getTags();
 
+  public ModuleKey getModuleKey() {
```
[the remark this answers is not in the document]
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java]: Replaced with a dedicated field.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionMetadata.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionMetadata.java:255]: I see. This is fine then.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleKey.java [hunk 5]
```diff
@@ -64,6 +65,19 @@ public final String toString() {
     return getName() + "@" + (getVersion().isEmpty() ? "_" : getVersion().toString());
   }
 
+  public static ModuleKey fromString(String s) throws ParseException {
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleKey.java:68]: is this still needed?

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleKey.java]: Thanks for catching this, remove the unused function.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleKey.java [hunk 6]
```diff
@@ -17,6 +17,7 @@
 
 import com.google.auto.value.AutoValue;
 import com.google.common.collect.ImmutableMap;
+import com.google.devtools.build.lib.bazel.bzlmod.Version.ParseException;
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleKey.java:20]: remove this too :)

**Review by Wyverald — APPROVED** [review]

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleKey.java (hunk 6 above)
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleKey.java]: Done :-)

**Comment by Pavank1992** [comment]
Hi @fmeum, As this PR is awaiting to merge, could you please fix the buildkite error? Thanks!

**Comment by fmeum** [comment]
> Hi @fmeum, As this PR is awaiting to merge, could you please fix the buildkite error? Thanks!

@Pavank1992 I am pretty sure they are not caused by this PR, but I triggered a rerun.

**Comment by Pavank1992** [comment]
Hi @fmeum, Yes, it needs to be added some comments. I added the `Bazel Authors` and `Apache License`, Version 2.0 (the "License") and syncing it. Thanks!

**Review by meteorcloudy — APPROVED** [review]

**Comment by fmeum** [comment]
@bazel-io flag

**Comment by Wyverald** [comment]
@bazel-io fork 6.2.0

**Pull request closed without merging**